File: desktop/src/test/unit/protocol/DriveCatalogContract.test.ts
import { describe, expect, it } from 'vitest';
import { DRIVE_BOOKS_PATH, DRIVE_SCOPE, canonicalBookName, canonicalBookPath, mergeCatalogVersion, reconcileLegacyReference, coverError, parseCanonicalBookName, type Lifecycle, type SyncErrorCode } from '$lib/shared/protocol/DriveCatalogContract';
const files = [{ fileId: 'legacy-id', name: 'old-name.epub', download: async () => new TextEncoder().encode('legacy') }, { fileId: 'canonical-id', name: 'book-1.epub', download: async () => new TextEncoder().encode('book') }];
describe('Drive catalog contract', () => {
  it('builds the canonical folder/file protocol', () => { expect(DRIVE_SCOPE).toContain('/drive.file'); expect(DRIVE_BOOKS_PATH).toBe('NextPage/Books'); expect(canonicalBookName('book-1', '.EPUB')).toBe('book-1.epub'); expect(canonicalBookPath('book-1', 'epub')).toBe('NextPage/Books/book-1.epub'); });
  it('parses canonical filenames into bookId + lowercase ext', () => {
    expect(parseCanonicalBookName('uuid.epub')).toEqual({ bookId: 'uuid', ext: 'epub' });
    expect(parseCanonicalBookName('uuid.PDF')).toEqual({ bookId: 'uuid', ext: 'pdf' });
    expect(parseCanonicalBookName('book-1.epub')).toEqual({ bookId: 'book-1', ext: 'epub' });
  });
  it('returns null for state files, empty, dotless, and trailing-dot names', () => {
    expect(parseCanonicalBookName('uuid_state.json')).toBeNull();
    expect(parseCanonicalBookName('')).toBeNull();
    expect(parseCanonicalBookName('uuid')).toBeNull();
    expect(parseCanonicalBookName('uuid.')).toBeNull();
  });
  it('reconciles by id, name, then independently verified hash', async () => { const r = await reconcileLegacyReference('legacy-id', files, 'book-1', 'epub', 'sha256:c49fea7425fa7f8699897a97c159c6690267d9003bb78c53fafa8fc15c325d84'); expect(r?.fileId).toBe('legacy-id'); expect(r?.fileName).toBe('book-1.epub'); expect(r?.canonicalPath).toBe('NextPage/Books/book-1.epub'); expect((await reconcileLegacyReference('missing', files, 'book-1', 'epub', null))?.fileId).toBe('canonical-id'); expect((await reconcileLegacyReference(null, files, 'book-1', 'epub', 'sha256:92719fe0cf8cd51592af31ee8a5736d79f7273777fa3f7b70bfe993a4cd32180'))?.fileId).toBe('canonical-id'); expect(await reconcileLegacyReference('canonical-id', files, 'book-1', 'epub', 'sha256:wrong')).toBeNull(); expect(await reconcileLegacyReference('legacy-id', [{ ...files[0], download: async () => { throw new Error('403'); } }], 'book-1', 'epub', 'sha256:c49fea7425fa7f8699897a97c159c6690267d9003bb78c53fafa8fc15c325d84')).toBeNull(); });
  it('ignores stale events and accepts equal/newer versions', () => { const current = { catalogVersion: 4, lifecycle: 'available' as Lifecycle }; expect(mergeCatalogVersion(current, { catalogVersion: 3, lifecycle: 'deleted' })).toBe(current); expect(mergeCatalogVersion(current, { catalogVersion: 4, lifecycle: 'deleted' }).lifecycle).toBe('deleted'); });
  it('exports COVER_FAILED as a stable SyncErrorCode', () => {
    // REQ-07: upload/invalid/missing cover failures must never block book import.
    const codes: SyncErrorCode[] = ['AUTH_REQUIRED', 'AUTH_EXPIRED', 'PERMISSION_DENIED', 'REMOTE_NOT_FOUND', 'HASH_MISMATCH', 'CONFLICT', 'UNAVAILABLE', 'COVER_FAILED'];
    expect(codes).toContain('COVER_FAILED');
    const err = coverError('corr-1', 'book-9');
    expect(err.code).toBe('COVER_FAILED');
    expect(err.retryable).toBe(true);
    expect(err.correlationId).toBe('corr-1');
    expect(err.bookId).toBe('book-9');
    const errNoBook = coverError('corr-2');
    expect(errNoBook.code).toBe('COVER_FAILED');
    expect(errNoBook.bookId).toBeUndefined();
  });
});

File: android/app/src/test/java/com/nextpage/data/remote/drive/DriveCatalogContractTest.kt
package com.nextpage.data.remote.drive

import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class DriveCatalogContractTest {
    private val legacy = LegacyRemoteCandidate("legacy-id", "old.epub") { "legacy".toByteArray() }
    private val canonical = LegacyRemoteCandidate("canonical-id", "book-1.epub") { "book".toByteArray() }
    @Test fun stableId_returnsCanonicalNameAndPath() = runBlocking {
        val r = resolve("legacy-id", listOf(legacy, canonical), "legacy")
        assertEquals("legacy-id", r?.fileId); assertEquals("book-1.epub", r?.fileName); assertEquals("NextPage/Books/book-1.epub", r?.canonicalPath)
    }
    @Test fun precedence_isIdThenNameThenHash() = runBlocking {
        assertEquals("legacy-id", resolve("legacy-id", listOf(legacy, canonical), "legacy")?.fileId)
        assertEquals("canonical-id", resolve(null, listOf(legacy, canonical), "book")?.fileId)
        assertEquals("legacy-id", resolve("missing", listOf(legacy), "legacy")?.fileId)
    }
    @Test fun unavailableSafe_returnsNullForMismatchOrDownloadFailure() = runBlocking {
        assertNull(resolve("legacy-id", listOf(legacy), "wrong")); val denied = LegacyRemoteCandidate("denied", "book-1.epub") { error("403") }
        assertNull(resolve("denied", listOf(denied), "book"))
    }
    @Test fun catalogRowWinner_isDeterministicByVersionThenUpdatedAtThenId() {
        fun row(id: String, v: Long, ts: String) = CatalogRow("u", id, "epub", "h", Lifecycle.AVAILABLE, null, v) to ts
        // Higher version wins.
        val (older, olderTs) = row("same", 3, "2025-01-01T00:00:00Z")
        val (newer, newerTs) = row("same", 4, "2025-06-01T00:00:00Z")
        assertEquals(4, catalogRowWinner(older, newer) { newerTs }.catalogVersion)
        // Equal version: newer updatedAt wins (row b returned).
        val (a, aTs) = row("same", 4, "2025-01-01T00:00:00Z")
        val (b, bTs) = row("same", 4, "2025-06-01T00:00:00Z")
        assertEquals("same", catalogRowWinner(a, b) { bTs }.bookId)
        assertEquals(4, catalogRowWinner(a, b) { bTs }.catalogVersion)
        // Equal version and updatedAt: greater id wins.
        val (c1, c1Ts) = row("book-a", 4, "2025-06-01T00:00:00Z")
        val (c2, c2Ts) = row("book-b", 4, "2025-06-01T00:00:00Z")
        assertEquals("book-b", catalogRowWinner(c1, c2) { c2Ts }.bookId)
    }
    @Test fun redactLogLine_stripsTokensJwtsAndHashesButKeepsCorrelation() {
        val jwt = "eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiIxMjMifQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c"
        val line = "sync token=secret123 authorization=Bearer $jwt hash=0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef correlation=abc-123"
        val redacted = redactLogLine(line)
        assertTrue(redacted.contains("token=[REDACTED]"))
        assertTrue(redacted.contains("authorization=[REDACTED]"))
        assertTrue(redacted.contains("[JWT_REDACTED]"))
        assertTrue(redacted.contains("[HASH_REDACTED]"))
        assertTrue(redacted.contains("correlation=abc-123"))
        assertFalse(redacted.contains("secret123"))
    }
    @Test fun syncErrorCodes_exportsStableCoverFailedCode() {
        // REQ-07: cover failures must never block book import and must be typed COVER_FAILED.
        assertEquals("COVER_FAILED", SyncErrorCodes.COVER_FAILED)
        val expected = setOf(
            "AUTH_REQUIRED", "AUTH_EXPIRED", "PERMISSION_DENIED", "REMOTE_NOT_FOUND",
            "HASH_MISMATCH", "CONFLICT", "UNAVAILABLE", "COVER_FAILED"
        )
        assertEquals(expected, setOf(
            SyncErrorCodes.AUTH_REQUIRED, SyncErrorCodes.AUTH_EXPIRED, SyncErrorCodes.PERMISSION_DENIED,
            SyncErrorCodes.REMOTE_NOT_FOUND, SyncErrorCodes.HASH_MISMATCH, SyncErrorCodes.CONFLICT,
            SyncErrorCodes.UNAVAILABLE, SyncErrorCodes.COVER_FAILED
        ))
    }
    @Test fun coverFailureError_mapsToStableCodeAndIsRetryable() {
        val err = coverFailureError("corr-1", "book-9")
        assertEquals("COVER_FAILED", err.code)
        assertTrue(err.retryable)
        assertEquals("corr-1", err.correlationId)
        assertEquals("book-9", err.bookId)
        val errNoBook = coverFailureError("corr-2")
        assertEquals("COVER_FAILED", errNoBook.code)
        assertNull(errNoBook.bookId)
    }
    private suspend fun resolve(id: String?, files: List<LegacyRemoteCandidate>, value: String) = DriveCatalogContract.reconcileLegacyReference(id, files, "book-1", "epub", "sha256:${sha256(value)}")
    private fun sha256(value: String) = java.security.MessageDigest.getInstance("SHA-256").digest(value.toByteArray()).joinToString("") { "%02x".format(it) }
}
